components/sbm-openrewrite/src/main/java/org/springframework/sbm/maven/UpgradeUnmanagedSpringProject.java

```diff
@@ -0,0 +1,210 @@
+                        if (isParentTag()) {
+
+                            Optional<Xml.Tag> artifactId = tag.getChild(ARTIFACT_ID);
+
+                            if (artifactId.isPresent()) {
+                                Optional<String> artifactIdValue = artifactId.get().getValue();
+                                if (artifactIdValue.isPresent()) {
+                                    if (artifactIdValue.get().equals(SPRING_BOOT_STARTER_PARENT)) {
+                                        validForFurtherReview = false;
+                                    }
+                                }
+                            }
+                        }
+                        if (isManagedDependencyTag(SPRINGBOOT_GROUP, SPRING_BOOT_DEPENDENCIES)) {
+                            validForFurtherReview = false;
+                        }
+                        return super.visitTag(tag, executionContext);
+                    }
+                }.visit(document, 0);
+
+                return super.visitDocument(document, executionContext);
+            }
+
+            @Override
+            public Xml.Tag visitTag(Xml.Tag tag, ExecutionContext executionContext) {
+                Xml.Tag resultTag = super.visitTag(tag, executionContext);
+
+                if (validForFurtherReview) {
+                    if (isManagedDependencyTag()) {
+
+                        ResolvedManagedDependency managedDependency = findManagedDependency(resultTag);
+
+                        if ((managedDependency != null) && managedDependency.getGroupId().equals(SPRINGBOOT_GROUP)
+                                && satisfiesOldVersionPattern(managedDependency.getVersion())) {
+                            return applyThisRecipe(resultTag);
+                        }
+                    }
+
+                    if (isDependencyTag()) {
+                        ResolvedDependency dependency = findDependency(resultTag);
+                        if ((dependency != null) && dependency.getGroupId().equals(SPRINGBOOT_GROUP)
+                                && satisfiesOldVersionPattern(dependency.getVersion())) {
+                            return applyThisRecipe(resultTag);
+                        }
+                    }
+                }
+
+                return resultTag;
+            }
+
+            @NotNull
+            private Xml.Tag applyThisRecipe(Xml.Tag resultTag) {
+                return resultTag.withMarkers(resultTag.getMarkers().addIfAbsent(new SearchResult(UUID.randomUUID(), "SpringBoot dependency")));
+            }
+
+            private boolean satisfiesOldVersionPattern(String version) {
+                return oldVersionPattern.matcher(version).matches();
+            }
+        };
+    }
+
+    @Override
+    public String getDisplayName() {
+        return "Upgrade unmanaged spring project";
+    }
+
+    public synchronized Map<String, String> getDependenciesMap() {
```

Comment:
getDependenciesMap() is singleton, designed to avoid downloading all SpringBoot dependencies multiple times. I prefer avoid that multiple threads call this method in parallel not only because of thread unsafe HashMap, but also because of downloading and resolving whole SpringBoot dependencies. I am not sure if it will work safe in parallel. It is not problem for now, because we use single thread from CLI, but can be potential issue after exposing remote API